pageserver/src/tenant.rs

```diff
@@ -1639,64 +1644,88 @@ impl Tenant {
+            // Spawn gc and compaction loops. The loops will shut themselves
+            // down when they notice that the tenant is inactive.
+            tasks::start_background_loops(self.tenant_id);
 
-                    let mut activated_timelines = 0;
+            let mut activated_timelines = 0;
 
-                    for timeline in not_broken_timelines {
-                        timeline.activate(broker_client.clone(), ctx);
-                        activated_timelines += 1;
-                    }
+            for timeline in not_broken_timelines {
+                timeline.activate(broker_client.clone(), ctx);
+                activated_timelines += 1;
+            }
 
-                    let elapsed = self.loading_started_at.elapsed();
-                    let total_timelines = timelines_accessor.len();
+            self.state.send_modify(move |current_state| {
+                assert!(
+                    *current_state == TenantState::Activating,
+                    "set_stopping and set_broken wait for us to leave Activating state",
+                );
+                *current_state = TenantState::Active;
 
-                    // log a lot of stuff, because some tenants sometimes suffer from user-visible
-                    // times to activate. see https://github.com/neondatabase/neon/issues/4025
-                    info!(
-                        since_creation_millis = elapsed.as_millis(),
-                        tenant_id = %self.tenant_id,
-                        activated_timelines,
-                        total_timelines,
-                        post_state = <&'static str>::from(&*current_state),
-                        "activation attempt finished"
-                    );
-                }
-            }
-        });
-        result
+                let elapsed = self.loading_started_at.elapsed();
+                let total_timelines = timelines_accessor.len();
+
+                // log a lot of stuff, because some tenants sometimes suffer from user-visible
+                // times to activate. see https://github.com/neondatabase/neon/issues/4025
+                info!(
+                    since_creation_millis = elapsed.as_millis(),
+                    tenant_id = %self.tenant_id,
+                    activated_timelines,
+                    total_timelines,
+                    post_state = <&'static str>::from(&*current_state),
+                    "activation attempt finished"
+                );
+            });
+        };
+        Ok(())
     }
 
-    /// Change tenant status to Stopping, to mark that it is being shut down
-    pub fn set_stopping(&self) {
+    /// Change tenant status to Stopping, to mark that it is being shut down.
+    ///
+    /// This function is not cancel-safe!
+    pub async fn set_stopping(&self) {
+        // Get the rx before checking state inside send_if_modified.
+        // This way, when we later rx.changed().await, we won't have missed
+        // any state changes.
```

Comment:
I'd prefer to drop this whole comment... It might even be outdated now?

No guaratees for observing every version with watch I think, ever, it could still hop multiple times ahead.

Might be worthwhile to refactor this `wait_for` (my suggestion below) out, and there add comments about why is this sane... I tried to come up with the state regex but it is not trivial to get it right in the time I have right now.

Reply:
Dropped the comment & switched to wait_for in fe4ef121b6c4ffeaf9175001c463cfdc598164d1